JASP-Engine/JASP/R/classicalmetaanalysis.R

```diff
@@ -0,0 +1,784 @@
+  #  dd = function(options, ...){x=list(...); options[names(x)] = x; options}
+  
+  # eval(parse(text = gsub("\\s+\\#","",readLines(pipe('pbpaste')))))
+  # dataset = metafor::escalc(measure="RR", ai=tpos, bi=tneg, ci=cpos, di=cneg, data=metafor::dat.bcg)
+  # dataset = transform(dataset, sei = vi^0.5)
+  #********
+  
+  #### Compute results of the analysis
+  
+  ## Get data. ####
+  
+  effsizeName     <- options$effectsize
+  stderrName      <- options$stderr
+  weightsName     <- options$wlsWeights
+  studyLabelName  <- options$studyLabels
+  
+  effsizeName <- unlist(options$dependent)
+  stderrName  <- unlist(options$wlsWeight)
+  
+  covarNames <- if (length(options$covariates) > 0) unlist(options$covariates)
+  factNames  <- if (length(options$factors) > 0) unlist(options$factors)
+
+  list.variables    <- Filter(function(s) s != "", c(effsizeName, covarNames))
+  numeric.variables <- Filter(function(s) s != "", c(effsizeName, covarNames, stderrName))
+  factor.variables  <- Filter(function(s) s != "", c(factNames, studyLabelName))
+  
+  
+  if (is.null(dataset)) {
+    .readDS <- switch(perform, run = .readDataSetToEnd, .readDataSetHeader)
+    dataset <- .readDS(
+      columns.as.factor  = factor.variables,
+      columns.as.numeric = numeric.variables,
+      exclude.na.listwise = c()	# let metafor::rma deal with NA's
+    )
+    
+    .hasErrors(dataset, perform, type=c("infinity", "observations"),
+      all.target=numeric.variables, observations.amount="< 2",
+      exitAnalysisIfErrors=TRUE)
+  }
+  
+  method = switch(options$method, `Fixed Effects` = "FE", `Maximum Likelihood` = "ML",
+      `Restricted ML` = "REML", `DerSimonian-Laird` = "DL", `Hedges` = "HE",
+      `Hunter-Schmidt` = "HS", `Sidik-Jonkman` = "SJ", `Empirical Bayes` = "EB",
+      `Paule-Mandel` = "PM", "REML")
+
+  can.run = all(c(effsizeName, stderrName) != "")
+  
+  ## Run the analysis. ####
+  if (is.null(rma.fit)) {
+    rma.fit <- structure(list('b' = numeric(),'se' = numeric(),'ci.lb' = numeric(),'ci.ub' = numeric(), 
+                    'zval' = numeric(),'pval' = numeric()), class = c("dummy", "rma"))
+    
+    if (run && can.run) {
+      
+      .vmodelTerms = options$modelTerms
+      formula.rhs <- as.formula(as.modelTerms(.vmodelTerms))
+      if (is.null(formula.rhs)) 
+        formula.rhs = ~1
+      if (!options$includeConstant) 
+        formula.rhs = update(formula.rhs, ~ . + 0)
+      if( identical(formula.rhs, ~ 1 - 1) )
+        .quitAnalysis("The model should contain at least one predictor or an intercept.")
+
+      rma.fit <- tryCatch(
+        metafor::rma(
+          yi = get(b64(effsizeName)), sei = get(b64(stderrName)), data = dataset,
+          method=method, mods = b64(formula.rhs), test = options$test,
+          slab = if(options$studyLabel != "") get(b64(options$studyLabels)),
+          level = options$regressionCoefficientsConfidenceIntervalsInterval
+        ), error = .quitAnalysis)
```

Comment:
In general .quitAnalysis is used when we expect the analysis to abort.
In these cases we can give a clear, informative error message.

I don't see the added benefit of this tryCatch, because it will still be an exception, not an expected error.